docs: new variants per environment (#2590)
Unleash/unleash: website/docs/reference/feature-toggle-variants.md

---
title: Feature Toggle Variants
---

> :::info Availability
>   **Feature toggle variants** were introduced in Unleash 3.2.
>   **Per-environment variants** were introduced in Unleash 4.21.
> :::

Reviewer: ```suggestion
:::info Availability

**Feature toggle variants** were first introduced in Unleash 3.2.

The option to have **different variant configurations in each environment** was introduced in Unleash 4.21.

:::
```

Author: I like it better, but I'm not sure I'd call it an option, because it's not optional. You have to deal with different variants configurations in each environment

Reviewer: Ah, that's a very good point. I mean, technically, you can have the same variants for all environments still: you just need to keep them in sync. But I agree that it's confusing and potentially misleading 😅 

How about this?

```suggestion
:::info Availability

**Feature toggle variants** were first introduced in Unleash 3.2.
In Unleash 4.21, variants were updated to be **environment-dependent**, meaning the same feature could have different variant configurations in different environments.

:::
```

Every toggle in Unleash can have something called _variants_. Where _feature toggles_ allow you to decide which users get access to a feature, _toggle variants_ allow you to further split those users into segments. Say, for instance, that you're testing out a new feature, such as an alternate sign-up form. The feature toggle would expose the form only to a select group of users. The variants could decide whether the user sees a blue or a green submit button on that form.

Variants facilitate A/B testing and experimentation by letting you create controlled and measurable experiments. Check [our blog post on using Unleash for A/B/n experiments](https://www.getunleash.io/blog/a-b-n-experiments-in-3-simple-steps) for some more insights into how you can set it up.

## What are variants?

> Whenever you create a feature toggle, you can assign it any number of _variants_. This is commonly done in cases where you want to serve your users different versions of a feature to see which performs better. You can assign different variants in each environment.

Reviewer: ```suggestion
Whenever you create a feature toggle, you can assign it any number of _variants_. This is commonly done in cases where you want to serve your users different versions of a feature to see which performs better. A feature can have different variants in each of its environments.
```

A variant has four components that define it:
- a **name**:

>     This must be unique among the toggle's variants in an environment. When working with a toggle with variants in a client, you will typically use the variant's name to find out which variant it is.

Reviewer: ```suggestion
    This must be unique among the toggle's variants in the specified environment. When working with a feature with variants in a client, you will typically use the variant's name to find out which variant it is.
```

---

> <div class="alert alert--info" role="alert">
>   <em>Feature toggle variants</em> were introduced in <em>Unleash v3.2.0</em>.
>   <em>Feature toggle variants</em> are attached to environments since <em>Unleash v4.19.0</em>.
> </div>
> <br/>

Reviewer: This should be rewritten as an admonition. Something like:

```suggestion
:::info Availability

**Feature toggle variants** were introduced in Unleash 3.2.0.

**Per-environment variants** (link to a description?) were introduced in Unleash 4.x.

:::
```

## What are variants?

> Whenever you create a feature toggle, you can assign it any number of _variants_ which can be different depending on the enviroment. This is commonly done in cases where you want to serve your users different versions of a feature to see which performs better.

Reviewer: ```suggestion
Whenever you create a feature toggle, you can assign it any number of _variants_. This is commonly done in cases where you want to serve your users different versions of a feature to see which performs better. You can assign different variants in each environment.
```
